### PR TITLE
@W-8461061 dragAndDrop and args type for element parameter

### PR DESCRIPTION
**Pull request by lizaiv77**
- Draggable interface
```java
public interface Draggable extends BasicElement {

  /**
   * drag and drop an element to the target location
   *
   * @param target location where to drop
   * @param holdDurationSec after clicking, hold for few seconds (emulates real user)
   */
  void dragAndDrop(BasicElement target, int holdDurationSec);

  /**
   * drag and drop an element to the target location
   *
   * @param target location where to drop
   */
  void dragAndDrop(BasicElement target);

  /**
   * drag and drop an element by offset
   *
   * @param xOffset horizontal offset
   * @param yOffset vertical offset
   * @param holdDurationSec after clicking, hold for few seconds (emulates real user)
   */
  void dragAndDropByOffset(int xOffset, int yOffset, int holdDurationSec);

  /**
   * drag and drop an element by offset
   *
   * @param xOffset horizontal offset
   * @param yOffset vertical offset
   */
  void dragAndDropByOffset(int xOffset, int yOffset);
}
```
- Element.dragAndDrop
```java
void dragAndDrop(DragAndDropOptions options);
interface DragAndDropOptions {

    /**
     * Duration of hold
     *
     * @return null by default
     */
    default Duration getHoldDuration() {
      return null;
    }

    /**
     * Target position for drop, as an element
     *
     * @return null by default
     */
    default Element getTargetElement() {
      return null;
    }

    /**
     * Target position for drop, as offset coordinates
     *
     * @return null by default
     */
    default Point getOffset() {
      return null;
    }
}
```
- new args type to pass an element as a parameter
- support for different number of parameters

______________________

Examples of dragAndDrop usage in compose: 
```json
{
"methods": [
    {
      "name": "composeDragAndDropElement",
      "compose": [
        {
          "element" : "first",
          "apply": "dragAndDrop",
          "args": [
            {
              "type": "element",
              "name": "elementArg"
            }
          ]
        }
      ]
    },
    {
      "name": "composeDragAndDropElementLiteralWithDuration",
      "compose": [
        {
          "element" : "first",
          "apply": "dragAndDrop",
          "args": [
            {
              "value": {
                "element" : "second"
              }
            },
            {
              "value": 2
            }
          ]
        }
      ]
    },
    {
      "name": "composeDragAndDropWithOffset",
      "compose": [
        {
          "element" : "simplePublic",
          "apply": "dragAndDrop",
          "args": [
            {
              "type": "number",
              "name": "x"
            },
            {
              "type": "number",
              "name": "y"
            }
          ]
        }
      ]
    },
    {

      "name": "composeDragAndDropWithOffsetDuration",
      "compose": [
        {
          "element" : "simplePublic",
          "apply": "dragAndDrop",
          "args": [
            {
              "value": 1
            },
            {
              "value": 2
            },
            {
              "type": "number",
              "name": "duration"
            }
          ]
        }
      ]
    }
  ]
}
```

### CODE REVIEW

**Comment by lizaiv77** [comment]
@jimevanssfdc PR is ready for review. it was green yesterday but pipeline was not triggered today. I'll try and push dummy commits to trigger it later

File: utam-core/src/main/java/utam/core/declarative/representation/MethodParameter.java [hunk 1]
```diff
@@ -7,21 +7,52 @@
...
 public interface MethodParameter {
 
+  /**
+   * check if parameter is literal and can be skipped in method signature declaration
+   *
+   * @return true if passed by value AKA literal
+   */
   boolean isLiteral();
 
-  boolean isSelectorArgument();
```
Comment (lizaiv77) [inline utam-core/src/main/java/utam/core/declarative/representation/MethodParameter.java:20]: @jimevanssfdc it used to be only selector that would have nested args, now element as well. hence replaced it with new method getNestedParameters()

**Review by jimevanssfdc — APPROVED** [review]

**Comment by lizaiv77** [comment]
@jimevanssfdc in [last commit ](https://github.com/salesforce/utam-java/pull/58/commits/9f5f54fb910138d0f6e6ef91d613ee796dd9b464) I added 4 different possible methods for drag and drop - (by element or coordinates) X (with/without duration), examples of usages are [here](https://github.com/salesforce/utam-java/blob/9f5f54fb910138d0f6e6ef91d613ee796dd9b464/utam-compiler/src/test/resources/composeDragAndDrop.json) 

Draggable now looks like this
```java
public interface Draggable extends BasicElement {

  /**
   * drag and drop an element to the target location
   *
   * @param target location where to drop
   * @param holdDurationSec after clicking, hold for few seconds (emulates real user)              
   */
  void dragAndDrop(BasicElement target, int holdDurationSec);

  /**
   * drag and drop an element to the target location
   *
   * @param target location where to drop
   */
  void dragAndDrop(BasicElement target);

  /**
   * drag and drop an element by offset
   *
   * @param xOffset horizontal offset
   * @param yOffset vertical offset
   * @param holdDurationSec after clicking, hold for few seconds (emulates real user)               
   */
  void dragAndDrop(int xOffset, int yOffset, int holdDurationSec);

  /**
   * drag and drop an element by offset
   *
   * @param xOffset horizontal offset
   * @param yOffset vertical offset
   */
  void dragAndDrop(int xOffset, int yOffset);
}
```

and Element interface:
```java
 void dragAndDrop(Driver driver, DragAndDropOptions options);

 interface DragAndDropOptions {

    default Duration getHoldDuration() {
      return null;
    }

    default Element getTargetElement() {
      return null;
    }

    default CoordinatesOffset getOffset() {
      return null;
    }

  }

  interface CoordinatesOffset {

    int getX();

    int getY();
  }
```

File: utam-core/src/main/java/utam/core/element/Draggable.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ */
+package utam.core.element;
+
+/**
+ * interaction methods for an element that can be "dragged"
+ *
+ * @author elizaveta.ivanova
+ * @since 236
+ */
+public interface Draggable extends BasicElement {
```
Comment (lizaiv77) [inline utam-core/src/main/java/utam/core/element/Draggable.java:16]: @jimevanssfdc pls check if you're ok with API

File: utam-core/src/main/java/utam/core/element/Element.java [hunk 3]
```diff
@@ -19,56 +20,154 @@
...
+   */
   boolean flickItems(GestureDirection direction);
 
+  /**
+   * drag and drop an element to the target location
+   *
+   * @param options location where to drop as element or coordinates offset and optional hold
+   *                duration
+   */
+  void dragAndDrop(DragAndDropOptions options);
```
Comment (lizaiv77) [inline utam-core/src/main/java/utam/core/element/Element.java:169]: @jimevanssfdc pls check if you're ok with API for dragAndDrop, mainly passing options as a single type, DragAndDropOptions is declared below

File: utam-core/src/main/java/utam/core/element/Element.java [hunk 4]
```diff
@@ -86,4 +185,40 @@
...
+    default Element getTargetElement() {
+      return null;
+    }
+
+    /**
+     * Target position for drop, as offset coordinates
+     *
+     * @return null by default
+     */
+    default Point getOffset() {
```
Comment (lizaiv77) [inline utam-core/src/main/java/utam/core/element/Element.java:219]: @jimevanssfdc 
is it ok to use Java.Point here? we don't want to use Selenium Point as this is Selenium agnostic wrapper for WebElement

File: utam-core/src/main/java/utam/core/element/Draggable.java (hunk 2 above)
Comment (jimevanssfdc) [reply utam-core/src/main/java/utam/core/element/Draggable.java:16]: Looks exactly right to me.

File: utam-core/src/main/java/utam/core/element/Element.java (hunk 4 above)
Comment (jimevanssfdc) [reply utam-core/src/main/java/utam/core/element/Element.java:219]: Let's create our own data structure. AWT may bring other baggage that we don't really want to have to deal with going forward. Or, if everything is just going off of an Options class anyway, do we really need it?

**Review by jimevanssfdc — CHANGES_REQUESTED** [review]

File: utam-core/src/main/java/utam/core/element/Element.java (hunk 4 above)
Comment (lizaiv77) [reply utam-core/src/main/java/utam/core/element/Element.java]: added to method getXOffset and getYOffset and moved this interface outside

File: utam-core/src/main/java/utam/core/element/DragAndDropOptions.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+import java.time.Duration;
+import utam.core.framework.UtamCoreError;
+
+/**
+ * options for a drag and drop actions: element, offset and duration.
+ *
+ * @since 236
+ * @author elizaveta.ivanova
+ */
+public interface DragAndDropOptions {
```
Comment (lizaiv77) [inline utam-core/src/main/java/utam/core/element/DragAndDropOptions.java:21]: @jimevanssfdc moved options here to separate location + replaced Point with 2 getters for offset

**Review by jimevanssfdc — APPROVED** [review]
LGTM! Good work.

**Pull request merged**